fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeContext.java

```diff
@@ -0,0 +1,248 @@
+                    Preconditions.checkState(false);
+                }
+                ColumnRefOperator useStringRef = useStringRefs.get(0);
+                ColumnRefOperator useDictRef = stringRefToDictRefMap.get(useStringRef);
+                // return type is dict
+                ScalarOperator dictExpr = new DictMappingOperator(useDictRef, stringDefineExpr.clone(), dictRef.getType());
+                dictRefToDefineExprMap.put(dictRef, dictExpr);
+            }
+        }
+    }
+
+    private void rewriteGlobalDict() {
+        GlobalDictRewriter dictRewriter = new GlobalDictRewriter();
+        for (Integer stringId : stringRefToDefineExprMap.keySet()) {
+            if (stringRefToDicts.containsKey(stringId)) {
+                continue;
+            }
+
+            // rewrite global dict expression
+            // rewrite to dictMapping(originDictColumn, fold_expression)
+            // e.g. A : upper(B)
+            //      B : lower(C)
+            // decode A: dictMapping(B, upper(B)) -> dictMapping(C, upper(lower(C)))
+            ColumnRefOperator dictRef = stringRefToDictRefMap.get(factory.getColumnRef(stringId));
+            ScalarOperator stringDefineExpr = dictRefToDefineExprMap.get(dictRef);
+            globalDictsExpr.put(dictRef.getId(), stringDefineExpr.accept(dictRewriter, null));
+        }
+    }
+
+    private void rewriteStringExpressions() {
+        // rewrite string expression
+        for (Integer stringId : stringExprsMap.keySet()) {
+            ColumnRefOperator stringRef = factory.getColumnRef(stringId);
+            ColumnRefOperator dictRef = stringRefToDictRefMap.get(stringRef);
+            for (ScalarOperator stringExpr : stringExprsMap.getOrDefault(stringId, Collections.emptyList())) {
+                if (stringExprToDictExprMap.containsKey(stringExpr)) {
+                    continue;
+                }
+                // return type is string, different as define expression
+                ScalarOperator dictExpr =
+                        new DictMappingOperator(dictRef, stringExpr.clone(), stringExpr.getType());
+                stringExprToDictExprMap.put(stringExpr, dictExpr);
+            }
+        }
+    }
+
+    private void rewriteStringAggregations() {
+        // rewrite string aggregate expression
+        FunctionRewriter rewriter = new FunctionRewriter();
+        for (CallOperator aggFn : stringAggregateExprs) {
+            CallOperator new1stAggFn = (CallOperator) (aggFn.accept(rewriter, null));
+            stringExprToDictExprMap.put(aggFn, new1stAggFn);
+        }
+    }
+
+    // create a new dictionary column and assign the same property except for the type and column id
+    // the input column maybe a dictionary column or a string column
+    private ColumnRefOperator createNewDictColumn(ColumnRefOperator column) {
+        if (column.getType().isArrayType()) {
+            return factory.create(column.getName(), Type.ARRAY_INT, column.isNullable());
```

Comment:
how about array_samllint?  In the future, the maximum low cardinality will be 65536

Reply:
I'm think fine, but all implement of BE is INT, change to samllint I think has compatibility issue